test/e2e/ingress_test.go

```diff
@@ -0,0 +1,277 @@
+		}
+	}
+
+	// Remove the "bar" consumingUser from the "foo" componentRoute and check that the roleBinding is updated
+	ingress.Status = configv1.IngressStatus{
+		ComponentRoutes: []configv1.ComponentRouteStatus{
+			{
+				Namespace: "default",
+				Name:      "foo",
+				ConsumingUsers: []string{
+					"foo",
+				},
+				DefaultHostname:  "foo.com",
+				CurrentHostnames: []string{"foo.com"},
+			},
+			{
+				Namespace: "default",
+				Name:      "bar",
+				ConsumingUsers: []string{
+					"bar",
+				},
+				DefaultHostname:  "bar.com",
+				CurrentHostnames: []string{"bar.com"},
+			},
+		},
+	}
+
+	if err := kclient.Status().Update(context.TODO(), ingress); err != nil {
+		t.Fatalf("failed to get ingress resource: %v", err)
+	}
+
+	for _, componentRoute := range ingress.Status.ComponentRoutes {
+		if err := pollForValidComponentRouteRoleBinding(t, componentRoute); err != nil {
+			t.Errorf("bad roleBinding: %v", err)
+		}
+	}
+
+	// Remove the `bar` componentRoute from the ingress spec and check that the related role and roleBinding are deleted
+	ingress.Spec.ComponentRoutes = []configv1.ComponentRouteSpec{
+		{
+			Namespace: "default",
+			Name:      "foo",
+			Hostname:  "www.testing.com",
+			ServingCertKeyPairSecret: configv1.SecretNameReference{
+				Name: "foo",
+			},
+		},
+	}
+
+	if err := wait.PollImmediate(1*time.Second, 10*time.Second, func() (bool, error) {
+		if err := kclient.Update(context.TODO(), ingress); err != nil {
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		t.Errorf("Failed to update ingress resource: %v", err)
+	}
+
+	// Make sure that the bar role and roleBinding are deleted
+	roleList := &rbacv1.RoleList{}
+	listOptions := []client.ListOption{
+		client.MatchingLabels{
+			componentRouteHashLabelKey: util.Hash("default/bar"),
+		},
+	}
+	if err := wait.PollImmediate(1*time.Second, 10*time.Second, func() (bool, error) {
+		if err := kclient.List(context.TODO(), roleList, listOptions...); err != nil {
+			return false, err
+		}
+		if len(roleList.Items) != 0 {
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		t.Errorf("role not deleted: %v", err)
+	}
+
+	roleBindingList := &rbacv1.RoleBindingList{}
+	if err := wait.PollImmediate(1*time.Second, 10*time.Second, func() (bool, error) {
+		if err := kclient.List(context.TODO(), roleBindingList, listOptions...); err != nil {
+			return false, err
+		}
+		if len(roleBindingList.Items) != 0 {
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		t.Errorf("role not deleted: %v", err)
+	}
+}
+
+func pollForValidComponentRouteRole(t *testing.T, componentRoute configv1.ComponentRouteSpec) error {
+	listOptions := []client.ListOption{
+		client.MatchingLabels{
+			componentRouteHashLabelKey: util.Hash(fmt.Sprintf("%s/%s", componentRoute.Namespace, componentRoute.Name)),
+		},
+	}
+
+	roleList := &rbacv1.RoleList{}
+	err := wait.PollImmediate(1*time.Second, 10*time.Second, func() (bool, error) {
```

Comment:
Would you say that 10 seconds is a reasonable timeout here? Wondering if this timeout (or some of the others in this test file) might be too aggressive.

Reply:
We can absolutely relax the timeout constraints (say to a minute) but I would be concerned if the controller ever takes more than a second or so to reconcile an event.

Reply:
I guess we can just keep on an eye on e2e operator test runs here?

Reply:
Latest run (https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/552/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1366461389803098112) shows some interesting timeouts that sparked my asking.

Reply:
I'll update the code to wait for a minute.

Reply:
Also - the test cases are failing because the [PR to update the ingresses.config.openshift.io CRD](https://github.com/openshift/cluster-config-operator/pull/187) has not merged, so the CRD's status does not include any component routes.